### ISSUE TITLE
Segfault when invoking exception handler on release builds

### ISSUE THREAD

**Issue by TJ09** [facebook/hhvm#7260]
### HHVM Version

Repros on 3.14.3 release build:
```
$ ./hhvm-3.14.3-release/bin/hhvm --version
HipHop VM 3.14.3 (rel)
Compiler: 1469739336_557442891
Repo schema: b2262b83e5a1ab11bf27e85c79af3e4ab53deb47
```

Does NOT repro on 3.14.3 debug build:
```
$ ./hhvm-3.14.3-debug/bin/hhvm --version
HipHop VM 3.14.3 (dbg)
Compiler: 1469694452_093991585
Repo schema: 42ef16eaf0b909c48f723899e3ea5a3a92d083ec
```

(Also does not repro on 3.12 LTS builds)

### Standalone code, or other way to reproduce the problem

```php
<?php

  function exception_handler($e) {
    echo "Caught exception\n";
  }

  set_exception_handler('exception_handler');

  throw new \Exception('test');
```

### Expected result
```
$ ./hhvm-3.14.3-debug/bin/hhvm segfault.php
Caught exception
```

### Actual result
```
$ ./hhvm-3.14.3-release/bin/hhvm segfault.php
Core dumped: Segmentation fault
Stack trace in /tmp/stacktrace.31091.log
```
(stacktrace.log is more or less blank)

From core dump:
```
Core was generated by `./hhvm-3.14.3-release/bin/hhvm segfault.php'.
Program terminated with signal 11, Segmentation fault.
#0  0x0000000001617437 in HPHP::ObjectData::invokeToString() ()
#0  0x0000000001617437 in HPHP::ObjectData::invokeToString() ()
#1  0x000000000181bdec in HPHP::Object::toString() const ()
#2  0x000000000184d597 in HPHP::handle_exception_helper(bool&, HPHP::ExecutionContext*, std::string&, HPHP::ContextOfException, bool&, bool) ()
#3  0x000000000184d8cc in HPHP::handle_exception(bool&, HPHP::ExecutionContext*, std::string&, HPHP::ContextOfException, bool&, bool) ()
#4  0x000000000358368a in HPHP::hphp_invoke(HPHP::ExecutionContext*, std::string const&, bool, HPHP::Array const&, HPHP::VRefParamValue const&, std::string const&, std::string const&, bool&, std::string&, bool, bool, bool) ()
#5  0x000000000184d9d3 in HPHP::hphp_invoke_simple(std::string const&, bool) ()
#6  0x000000000185fecf in HPHP::execute_program_impl(int, char**) ()
#7  0x0000000001863185 in HPHP::execute_program(int, char**) ()
#8  0x0000000000e6188b in main ()
```

**Comment by Orvid** [comment]
Which OS? I spent a painful amount of time debugging basically this same issue a while back (It was a massive pain, and is the result of a compiler bug). I didn't think the issue itself had made its way into any releases, but this makes it sound like it has.

We'll be putting out a security release on Monday, so if you can get me the OS's it happens on I can make sure the fix gets applied

**Comment by TJ09** [comment]
It's not any of the official releases (compiled myself), but it occurs on both Slackware 14.1 and 14.2 with GCC 4.8 and GCC 5.3. No custom flags were given to cmake.

I can see about trying to repro in a docker instance on a more supported OS, but I figured the differing compiler environments were enough that it wasn't likely to be a compiler issue.

**Comment by Orvid** [comment]
5.3 shouldn't have the issue, https://github.com/facebook/hhvm/commit/d239f8660e7c3bff35d619a500c63f4a1f10eae5 is where I re-enabled the bit that had been causing issues with all of 4.8...

**Comment by TJ09** [comment]
Oh wait, yep, the Slackware 14.2 build with GCC 5.3 doesn't have this issue, only the one built with GCC 4.8.2.

I was confusing this crash with #7231.

**Comment by Orvid** [comment]
I've cherry-picked the fix for this into the 3.14.4 release. See https://github.com/facebook/hhvm/commit/c1559a07632ee44c9dfb77e27a7c86524396ffae.
